**Scraping Business information from yelp with Selenium and Beautiful Soup. To get Started we need to install Selenium with Chrome driver**

In [ ]:
!pip install selenium requests_html
!pip install requests_html
!apt-get update 
!apt install chromium-chromedriver    #for google colab


     |████████████████████████████████| 133kB 18.7MB/s 
ERROR: datascience 0.10.6 has requirement folium==0.2.1, but you'll have folium 0.8.3 which is incompatible.
  Found existing installation: urllib3 1.26.4
    Uninstalling urllib3-1.26.4:
      Successfully uninstalled urllib3-1.26.4


Hit:1 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease
Ign:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease
Hit:3 http://security.ubuntu.com/ubuntu bionic-security InRelease
Hit:4 http://archive.ubuntu.com/ubuntu bionic InRelease
Ign:5 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease
Hit:6 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release
Hit:7 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release
Hit:8 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease
Hit:9 http://archive.ubuntu.com/ubuntu bionic-updates InRelease
Hit:10 http://archive.ubuntu.com/ubuntu bionic-backports InRelease
Hit:11 http://ppa.launchpad.net/cran/libgit2/ubuntu bionic InRelease
Hit:12 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic InRelease
Hit:13 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic 

**Initializing the chrome driver**

In [ ]:
# opening a webpage
import pandas as pd
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
                #add proxy for not getting block

chrome_options = webdriver.ChromeOptions()
chrome_options.page_load_strategy = 'eager'
chrome_options.add_argument('--headless')
chrome_options.add_argument('--no-sandbox')
chrome_options.add_argument('--disable-dev-shm-usage')
#chrome_options.add_argument('--proxy-server=%s' % PROXY)
browser =webdriver.Chrome('chromedriver',options=chrome_options)

**Searching and Generating the page links**

In [ ]:
all_page_urls = []     #to save the page urls
business_pages =[]     #to save the business page url

search_item = str(input('What do you want to search for? '))  #you can change what your wanna search for
location = str(input('Enter the location to search for: '))    #change in which loaction you want to search

base_url = "https://www.yelp.com/search?find_desc=" +search_item + "&find_loc="+location+"&start="  #main search page urls pattern

**Generating the main search page and pagination**

In [ ]:
def yelp_search_link():
  for i in range(0,200,10):       #change here for your requirement
    main_url = base_url+str(i)
    all_page_urls.append(all_page_urls)
    print(main_url)

yelp_search_link()   

https://www.yelp.com/search?find_desc=&find_loc=&start=0
https://www.yelp.com/search?find_desc=&find_loc=&start=10
https://www.yelp.com/search?find_desc=&find_loc=&start=20
https://www.yelp.com/search?find_desc=&find_loc=&start=30
https://www.yelp.com/search?find_desc=&find_loc=&start=40
https://www.yelp.com/search?find_desc=&find_loc=&start=50
https://www.yelp.com/search?find_desc=&find_loc=&start=60
https://www.yelp.com/search?find_desc=&find_loc=&start=70
https://www.yelp.com/search?find_desc=&find_loc=&start=80
https://www.yelp.com/search?find_desc=&find_loc=&start=90
https://www.yelp.com/search?find_desc=&find_loc=&start=100
https://www.yelp.com/search?find_desc=&find_loc=&start=110
https://www.yelp.com/search?find_desc=&find_loc=&start=120
https://www.yelp.com/search?find_desc=&find_loc=&start=130
https://www.yelp.com/search?find_desc=&find_loc=&start=140
https://www.yelp.com/search?find_desc=&find_loc=&start=150
https://www.yelp.com/search?find_desc=&find_loc=&start=160
https://

**We will only scrape the data from first page.**

**To extract the business information from each search page, we need to find the dedicated page links for all business listings. Each yelp search page contains 10 listing.**

In [ ]:
from random import randint
from time import sleep   
from bs4 import BeautifulSoup

def find_items_page():
    browser.get('https://www.yelp.com/search?find_desc=Hotel&find_loc=London&start=0')
    sleep(randint(1,10))
    soup = BeautifulSoup(browser.page_source, 'html.parser')
    sleep(randint(1,10))
    mains = soup.find_all('div', class_ = 'container__09f24__21w3G hoverable__09f24__2nTf3 margin-t3__09f24__5bM2Z margin-b3__09f24__1DQ9x padding-t3__09f24__-R_5x padding-r3__09f24__1pBFG padding-b3__09f24__1vW6j padding-l3__09f24__1yCJf border--top__09f24__8W8ca border--right__09f24__1u7Gt border--bottom__09f24__xdij8 border--left__09f24__rwKIa border-color--default__09f24__1eOdn')
    main_url = 'https://www.yelp.com'
    for main in mains:
      a_tag = main.find('a', class_ = 'css-166la90').get('href')
      sleep(randint(1,5))
      a_tag_formated = main_url + str(a_tag)
      business_pages.append(a_tag_formated)
      print(a_tag_formated)
find_items_page()

https://www.yelp.com/biz/the-hoxton-shoreditch-london?osq=Hotel
https://www.yelp.com/biz/london-marriott-hotel-park-lane-london?osq=Hotel
https://www.yelp.com/biz/hotel-41-london?osq=Hotel
https://www.yelp.com/biz/savoy-hotel-london?osq=Hotel
https://www.yelp.com/biz/the-ritz-london-london-2?osq=Hotel
https://www.yelp.com/biz/claridges-london-2?osq=Hotel
https://www.yelp.com/biz/the-dorchester-london?osq=Hotel
https://www.yelp.com/biz/corinthia-hotel-london-london-2?osq=Hotel
https://www.yelp.com/biz/the-ned-london?osq=Hotel
https://www.yelp.com/biz/jw-marriott-grosvenor-house-london-london?osq=Hotel


Now We have all the 10 links from first page.

In [ ]:
business_pages

['https://www.yelp.com/biz/the-hoxton-shoreditch-london?osq=Hotel',
 'https://www.yelp.com/biz/london-marriott-hotel-park-lane-london?osq=Hotel',
 'https://www.yelp.com/biz/hotel-41-london?osq=Hotel',
 'https://www.yelp.com/biz/savoy-hotel-london?osq=Hotel',
 'https://www.yelp.com/biz/the-ritz-london-london-2?osq=Hotel',
 'https://www.yelp.com/biz/claridges-london-2?osq=Hotel',
 'https://www.yelp.com/biz/the-dorchester-london?osq=Hotel',
 'https://www.yelp.com/biz/corinthia-hotel-london-london-2?osq=Hotel',
 'https://www.yelp.com/biz/the-ned-london?osq=Hotel',
 'https://www.yelp.com/biz/jw-marriott-grosvenor-house-london-london?osq=Hotel']

**Before you try to scrap a large number of pages, please be sure that you're using proxies. Otherwise, you will get banned for sending too many requests to the server. I have been banned from yelp. So be careful.**

In [ ]:

Names = []            #Name of the business profile
Reviews = []          #No of reviews recieved
Open_Hour = []        #Open hours
Price_range = []      #Price range
Address = []          #Address of the business
Websites = []         #Wbsites of the business
Phones = []           #Phone number of the business


def scrape_and_save():
  for urls in business_pages:
    browser.get(urls)
    sleep(randint(1,10))
    ss = BeautifulSoup(browser.page_source, 'html.parser')
    mainPage = ss.find('div', class_ = 'main-content-wrap main-content-wrap--full')
    sleep(randint(1,10))

    try:
        name = mainPage.find('h1', class_ = 'css-11q1g5y').get_text()
        Names.append(name)
    except AttributeError:
        name = 'NUll'
        Names.append(name)

    try:
        review = mainPage.find('span', class_ = 'css-bq71j2').get_text()
        Reviews.append(review)
    except AttributeError:
        review = 'Null'
        Reviews.append(review)

    try:
        price = mainPage.find('span', class_ = 'css-1xxismk').get_text()
        Price_range.append(price)
    except AttributeError:
        price = 'Null'
        Price_range.append(price)

    try:
        opens = mainPage.find('div', class_="display--inline-block__373c0__2de_K margin-r1-5__373c0__1Vie3 border-color--default__373c0__2oFDT").find_next('span', class_ = 'css-bq71j2').get_text()
        Open_Hour.append(opens)
    except AttributeError:
        opens = 'Null'
        Open_Hour.append(opens)

    try:
        address = mainPage.find('div', class_ = 'css-1vhakgw border--top__373c0__19Owr border-color--default__373c0__2oFDT').find_next('p', class_ = 'css-1h1j0y3').find_next('p', class_ = 'css-e81eai').get_text()
        Address.append(address)
    except AttributeError:
        address = 'Null'
        Address.append(address)
    except TypeError:
        address = 'Null'
        Address.append(address)

    try:
        website = mainPage.find('a', class_ = 'css-ac8spe').get_text()
        website = 'https://www.'+ str(website)
        Websites.append(website)
    except AttributeError:
        website = 'Null'
        Websites.append(website)

    try:
        phone = mainPage.find('div', class_ = 'stickySidebar__373c0__3PY1o border-color--default__373c0__2oFDT').find('p', class_ = 'css-1h1j0y3').find_next('p', class_ = 'css-1h1j0y3').get_text()
        Phones.append(phone)
    except AttributeError:
        phone = 'Null'
        Phones.append(phone)

    print('Scraping Compeleted',urls)

  df = pd.DataFrame({'Names':Names, 'Price Range': Price_range, 'Reviews': Reviews,
                  'Address':Address, 'Website': Websites,'Phone Number': Phones})  #making a pandas dataframe

  df.to_csv('Business Data.csv')   #Saving the data as csv
    
scrape_and_save()
    



Scraping Compeleted https://www.yelp.com/biz/the-hoxton-shoreditch-london?osq=Hotel
Scraping Compeleted https://www.yelp.com/biz/london-marriott-hotel-park-lane-london?osq=Hotel
Scraping Compeleted https://www.yelp.com/biz/hotel-41-london?osq=Hotel
Scraping Compeleted https://www.yelp.com/biz/savoy-hotel-london?osq=Hotel
Scraping Compeleted https://www.yelp.com/biz/the-ritz-london-london-2?osq=Hotel
Scraping Compeleted https://www.yelp.com/biz/claridges-london-2?osq=Hotel
Scraping Compeleted https://www.yelp.com/biz/the-dorchester-london?osq=Hotel
Scraping Compeleted https://www.yelp.com/biz/corinthia-hotel-london-london-2?osq=Hotel
Scraping Compeleted https://www.yelp.com/biz/the-ned-london?osq=Hotel
Scraping Compeleted https://www.yelp.com/biz/jw-marriott-grosvenor-house-london-london?osq=Hotel


**Scraping in this process may seem slow, but it is the safest option to scrap a large number of pages without getting banned.**

In [3]:
import pandas as pd
data = pd.read_csv('/content/Business Data.csv')
data

Unnamed: 0  ...   Phone Number
0           0  ...  020 7550 1000
1           1  ...  020 7493 7000
2           2  ...  020 7300 0041
3           3  ...  020 7836 4343
4           4  ...  020 7493 8181
5           5  ...  020 7629 8860
6           6  ...  020 7629 8888
7           7  ...  020 7930 8181
8           8  ...  020 3828 2000
9           9  ...  020 7499 6363

[10 rows x 7 columns]

**If you want to send all the requests synchronously then, you can use python's async requests feature. But without proxies using this method might get you into getting banned by the website. So, to avoid this, you can use proxies.**

**I have used grequests to send async requests for all the link(240).And after scraping more thand 160 I got banned from yelp.**

In [ ]:
pip install grequests

In [ ]:
from bs4 import BeautifulSoup
import grequests              #for async requests
import pandas as pd
import time


Generating page links

In [7]:
def get_urls():
    urls = []
    search_item = 'Hotel'           #you can change for your search
    location = 'London'             #change where you wanna search
    for x in range(0,200,10):
        urls.append("https://www.yelp.com/search?find_desc=" +search_item + "&find_loc="+location+"&start="+str(x))
    return urls
get_urls()

['https://www.yelp.com/search?find_desc=Hotel&find_loc=London&start=0',
 'https://www.yelp.com/search?find_desc=Hotel&find_loc=London&start=10',
 'https://www.yelp.com/search?find_desc=Hotel&find_loc=London&start=20',
 'https://www.yelp.com/search?find_desc=Hotel&find_loc=London&start=30',
 'https://www.yelp.com/search?find_desc=Hotel&find_loc=London&start=40',
 'https://www.yelp.com/search?find_desc=Hotel&find_loc=London&start=50',
 'https://www.yelp.com/search?find_desc=Hotel&find_loc=London&start=60',
 'https://www.yelp.com/search?find_desc=Hotel&find_loc=London&start=70',
 'https://www.yelp.com/search?find_desc=Hotel&find_loc=London&start=80',
 'https://www.yelp.com/search?find_desc=Hotel&find_loc=London&start=90',
 'https://www.yelp.com/search?find_desc=Hotel&find_loc=London&start=100',
 'https://www.yelp.com/search?find_desc=Hotel&find_loc=London&start=110',
 'https://www.yelp.com/search?find_desc=Hotel&find_loc=London&start=120',
 'https://www.yelp.com/search?find_desc=Hotel&fin

**Now we need to send async requests for all of these pages. This will give us the business page links from all 19 pages.**

In [ ]:
def get_data(urls):
  reqs = [grequests.get(link) for link in urls]
  time.sleep(0.15)
  resp = grequests.map(reqs, size=4)
  return resp

def parse(resp):
  business_page_links = []
  for r in resp:
      sp = BeautifulSoup(r.text, 'lxml')
      mains = sp.find_all('div', class_ = 'container__09f24__21w3G hoverable__09f24__2nTf3 margin-t3__09f24__5bM2Zmargin-b3__09f24__1DQ9x padding-t3__09f24__-R_5x padding-r3__09f24__1pBFG padding-b3__09f24__1vW6j padding-l3__09f24__1yCJfborder--top__09f24__8W8ca border--right__09f24__1u7Gt border--bottom__09f24__xdij8 border--left__09f24__rwKIa border-color--default__09f24__1eOdn')
      main_url = 'https://www.yelp.com'
      for main in mains:
          a_tag = main.find('a', class_ = 'css-166la90').get('href')
          a_tag_formated = main_url + str(a_tag)
          business_page_links.append(a_tag_formated)
          print('Added: ', a_tag_formated)
  return business_page_links


urls = get_urls()
resp = get_data(urls)
df = pd.DataFrame(parse(resp))
df.to_csv('Business page links.csv', index=False, header=Fals

In [9]:
links = open('/content/Business page links.csv').readlines()
links

['https://www.yelp.com/biz/the-hoxton-shoreditch-london?osq=Hotel\n',
 'https://www.yelp.com/biz/london-marriott-hotel-park-lane-london?osq=Hotel\n',
 'https://www.yelp.com/biz/hotel-41-london?osq=Hotel\n',
 'https://www.yelp.com/biz/savoy-hotel-london?osq=Hotel\n',
 'https://www.yelp.com/biz/the-ritz-london-london-2?osq=Hotel\n',
 'https://www.yelp.com/biz/the-dorchester-london?osq=Hotel\n',
 'https://www.yelp.com/biz/claridges-london-2?osq=Hotel\n',
 'https://www.yelp.com/biz/corinthia-hotel-london-london-2?osq=Hotel\n',
 'https://www.yelp.com/biz/the-ned-london?osq=Hotel\n',
 'https://www.yelp.com/biz/jw-marriott-grosvenor-house-london-london?osq=Hotel\n',
 'https://www.yelp.com/biz/the-landmark-london-london-2?osq=Hotel\n',
 'https://www.yelp.com/biz/hotel-indigo-london-3?osq=Hotel\n',
 'https://www.yelp.com/biz/the-kensington-hotel-london?osq=Hotel\n',
 'https://www.yelp.com/biz/four-seasons-hotel-london-4?osq=Hotel\n',
 'https://www.yelp.com/biz/mandarin-oriental-london-2?osq=Hot

**Now we will scrap the information from all of these links.**

In [ ]:
def get_data():
  reqs = [grequests.get(link) for link in (open('/content/Business page links.csv').readlines())]
  time.sleep(0.23)
  resp = grequests.map(reqs, size = 5)
  return resp

def parse(resp):
  Names = []
  Reviews = []
  Price_range = []
  Open_Hour = []
  Address = []
  Websites = []
  Phones = []
  for r in resp:
    soup = BeautifulSoup(r.text, 'lxml')
    time.sleep(0.22)
    mainPage = soup.find('div', class_ = 'main-content-wrap main-content-wrap--full')

    try:
        name = mainPage.find('h1', class_ = 'css-11q1g5y').get_text()
        Names.append(name)
    except AttributeError:
        name = 'NUll'
        Names.append(name)

    try:
        review = mainPage.find('span', class_ = 'css-bq71j2').get_text()
        Reviews.append(review)
    except AttributeError:
        review = 'Null'
        Reviews.append(review)

    try:
        price = mainPage.find('span', class_ = 'css-1xxismk').get_text()
        Price_range.append(price)
    except AttributeError:
        price = 'Null'
        Price_range.append(price)

    try:
        opens = mainPage.find('div', class_="display--inline-block__373c0__2de_K margin-r1-5__373c0__1Vie3 border-color--default__373c0__2oFDT").find_next('span', class_ = 'css-bq71j2').get_text()
        Open_Hour.append(opens)
    except AttributeError:
        opens = 'Null'
        Open_Hour.append(opens)

    try:
        address = mainPage.find('div', class_ = 'css-1vhakgw border--top__373c0__19Owr border-color--default__373c0__2oFDT').find_next('p', class_ = 'css-1h1j0y3').find_next('p', class_ = 'css-e81eai').get_text()
        Address.append(address)
    except AttributeError:
        address = 'Null'
        Address.append(address)
    except TypeError:
        address = 'Null'
        Address.append(address)

    try:
        website = mainPage.find('a', class_ = 'css-ac8spe').get_text()
        website = 'https://www.'+ str(website)
        Websites.append(website)
    except AttributeError:
        website = 'Null'
        Websites.append(website)
    try:
        phone = mainPage.find('div', class_ = 'stickySidebar__373c0__3PY1o border-color--default__373c0__2oFDT').find('p', class_ = 'css-1h1j0y3').find_next('p', class_ = 'css-1h1j0y3').get_text()
        Phones.append(phone)
    except AttributeError:
        phone = 'Null'
        Phones.append(phone)

    print(r)
    print('Added:', name)
  df = pd.DataFrame({'Names':Names, 'Price Range': Price_range, 'Open Hours':Open_Hour, 'Reviews': Reviews, 'Address':Address, 'Website': Websites,'Phone Number': Phones})
  df.to_csv('Business Data async.csv', index=False)


resp = get_data()
parse(resp)

In [10]:
data_async = pd.read_csv('/content/Business Data async.csv')
data_async

Names  ...   Phone Number
0            The Hoxton - Shoreditch  ...  020 7550 1000
1    London Marriott Hotel Park Lane  ...  020 7493 7000
2                           Hotel 41  ...  020 7300 0041
3                        Savoy Hotel  ...  020 7836 4343
4                    The Ritz London  ...  020 7493 8181
..                               ...  ...            ...
158                     The Draycott  ...  020 7730 6466
159                      The Rookery  ...  020 7336 0931
160          Best Western Palm Hotel  ...  020 8455 5220
161                  The Lodge Hotel  ...  020 8874 1598
162                      Byron Hotel  ...  020 7243 0987

[163 rows x 7 columns]

**Try all of these methods to scrap your desired information.**